docs/changing-the-ast.md

```diff
@@ -0,0 +1,66 @@
+---
+title: Changing the AST
+category: Compiler Internals
+categoryindex: 200
+index: 800
+---
+# Changing the AST
+
+Making changes to the AST is a common task when working on new F# compiler features or when working on developer tooling.  
+This document describes the process of making changes to the AST.
+
+The easiest way to modify the AST is to start with the type definitions in `SyntaxTree.fsi` and `SyntaxTree.fs` and then let the compiler guide you to the places where you need to make changes.
+Let's look at an example. Here, we want to add a new field `divRange` to the union case `Rational` of `SynRationalConst`:
+
+```fsharp
+type SynRationalConst =
+
+    // ...
+
+    | Rational of
+        numerator: int32 *
+        numeratorRange: range *
+        divRange: range *   // our new field
```

Comment:
Maybe this needs a little context as to why this is not part of `SyntaxTrivia`.
The pros of adding it to trivia:
- Having the additional information in a separate structure allows it to grow more easily over time. Adding new information to an existing trivia type won't disrupt most FCS consumers.
- It is clear that it is information that is not relevant for the compilation.

The cons of introducing trivia:
- It can be a bit excessive to introduce for a single field.
- The existing AST node might already contain fields that are historically more suited for trivia, but predate the `SyntaxTrivia` module.

In this case, adding one last range to the union case felt pretty reasonable.
But again, it is somewhat debatable. Maybe add something in the documentation that contributors should outweigh the pros and cons.